### PR TITLE
Mon 20969 - have the possibility to invert filters accepted hostgroups rejected hostgroups

### PR DESCRIPTION
**Pull request by sdepassio**
Filters added:
- rejected_hostgroups
- rejected_servicegroups
- rejected_bvs
- rejected_pollers
- rejected_authors

### CODE REVIEW

File: modules/centreon-stream-connectors-lib/sc_event.lua [hunk 1]
```diff
@@ -446,91 +446,133 @@ end
 function ScEvent:is_valid_hostgroup()
   self.event.cache.hostgroups = self.sc_broker:get_hostgroups(self.event.host_id)
 
-  -- return true if option is not set
-  if self.params.accepted_hostgroups == "" then
+  -- return true if options are not set or if both options are set
+  local accepted_hostgroups_isnotempty = self.params.accepted_hostgroups ~= ""
+  local rejected_hostgroups_isnotempty = self.params.rejected_hostgroups ~= ""
+  if (not accepted_hostgroups_isnotempty and not rejected_hostgroups_isnotempty) or (accepted_hostgroups_isnotempty and rejected_hostgroups_isnotempty) then
```
Comment (lucie-dubrunfaut) [inline modules/centreon-stream-connectors-lib/sc_event.lua:452]: I'm not sure but you are checking if : 
- only accepted hostgroup is set
- accepted host group and rejected host group are set together
but not if only rejected host group is set ?
Don't user may needs to only set rejected host group ?

File: modules/centreon-stream-connectors-lib/sc_event.lua [hunk 2]
```diff
@@ -663,17 +718,20 @@ function ScEvent:is_valid_bv()
 end
 
 --- find_bv_in_list: compare accepted BVs from parameters with the event BVs
--- @return accepted_name (string) the name of the first matching BV
+-- @return bv_name (string) the name of the first matching BV
```
Comment (tanguyvda) [inline modules/centreon-stream-connectors-lib/sc_event.lua:721]:
```suggestion
-- @param bvs_list (string) a coma separated list of BV name
-- @return bv_name (string) the name of the first matching BV
```

File: modules/centreon-stream-connectors-lib/sc_event.lua [hunk 3]
```diff
@@ -723,14 +793,17 @@ function ScEvent:is_valid_poller()
 end
 
 --- find_poller_in_list: compare accepted pollers from parameters with the event poller
--- @return accepted_name or false (string|boolean) the name of the first matching poller if found or false if not found
-function ScEvent:find_poller_in_list()
-  for _, accepted_name in ipairs(self.sc_common:split(self.params.accepted_pollers, ",")) do
-    if accepted_name == self.event.cache.poller then
-      return accepted_name
+-- @return poller_name or false (string|boolean) the name of the first matching poller if found or false if not found
```
Comment (tanguyvda) [inline modules/centreon-stream-connectors-lib/sc_event.lua:796]:
```suggestion
-- @param pollers_list (string) a coma separated list of poller name
-- @return poller_name or false (string|boolean) the name of the first matching poller if found or false if not found
```

File: modules/centreon-stream-connectors-lib/sc_event.lua [hunk 4]
```diff
@@ -977,13 +1058,16 @@ end
 
 --- find_author_in_list: compare accepted authors from parameters with the event author
 -- @return accepted_alias or false (string|boolean) the alias of the first matching author if found or false if not found
```
Comment (tanguyvda) [inline modules/centreon-stream-connectors-lib/sc_event.lua:1060]:
```suggestion
-- @param authors_list (string) a coma separeted list of author name
-- @return accepted_alias or false (string|boolean) the alias of the first matching author if found or false if not found
```

File: modules/centreon-stream-connectors-lib/sc_event.lua [hunk 5]
```diff
@@ -446,91 +446,133 @@ end
...
+    end
+    self.sc_logger:debug(debug_msg)
   end
 
   return true
 end
 
 --- find_hostgroup_in_list: compare accepted hostgroups from parameters with the event hostgroups
--- @return accepted_name (string) the name of the first matching hostgroup 
+-- @return hostgroup_name (string) the name of the first matching hostgroup
```
Comment (tanguyvda) [inline modules/centreon-stream-connectors-lib/sc_event.lua:495]:
```suggestion
-- @param hostgroups_list (string) a coma separated list of hostgroup name
-- @return hostgroup_name (string) the name of the first matching hostgroup
```

File: modules/centreon-stream-connectors-lib/sc_event.lua [hunk 6]
```diff
@@ -446,91 +446,133 @@ end
...
 end
 
 --- find_servicegroup_in_list: compare accepted servicegroups from parameters with the event servicegroups
--- @return accepted_name or false (string|boolean) the name of the first matching servicegroup if found or false if not found
-function ScEvent:find_servicegroup_in_list()
-  for _, accepted_name in ipairs(self.sc_common:split(self.params.accepted_servicegroups, ",")) do
-    for _, event_servicegroup in pairs(self.event.cache.servicegroups) do
-      if accepted_name == event_servicegroup.group_name then
-        return accepted_name
+-- @return servicegroup_name or false (string|boolean) the name of the first matching servicegroup if found or false if not found
```
Comment (tanguyvda) [inline modules/centreon-stream-connectors-lib/sc_event.lua:563]:
```suggestion
-- @param servicegroups_list (string) a coma separated list of servicegroup name
-- @return servicegroup_name or false (string|boolean) the name of the first matching servicegroup if found or false if not found
```

File: modules/docs/sc_event.md [hunk 7]
```diff
@@ -697,7 +699,13 @@ local result = test_event:is_valid_bv()
...
 ## find_hostgroup_in_list method
 
-The **find_hostgroup_in_list** method checks if one of the hostgroup in [**accepted_hostgroups**](sc_param.md#default-parameters) is linked to the host.
+The **find_hostgroup_in_list** method checks if one of the hostgroup in the hostgroups list parameter ([**accepted_hostgroups or rejected_hostgroups parameters**](sc_param.md#default-parameters)) is linked to the host.
+
+### find_hostgroup_in_list: parameters
+
+| parameter            | type   | optional | default value |
+| -------------------- | ------ | -------- | ------------- |
+| a list of hostgroups | string | no       |               |
```
Comment (tanguyvda) [inline modules/docs/sc_event.md:708]:
```suggestion
| a coma separated list of hostgroups | string | no       |               |
```

Comment (tanguyvda) [reply modules/docs/sc_event.md:708]: don't mind the fucked up indentation, it'll be fixed when we update the doc from code editor

File: modules/docs/sc_event.md [hunk 8]
```diff
@@ -712,19 +720,25 @@ The **find_hostgroup_in_list** method checks if one of the hostgroup in [**accep
...
 ## find_servicegroup_in_list method
 
-The **find_servicegroup_in_list** method checks if one of the servicegroup in [**accepted_servicegroups**](sc_param.md#default-parameters) is linked to the service.
+The **find_servicegroup_in_list** method checks if one of the servicegroup in the servicegroups list parameter ([**accepted_servicegroups or rejected_servicegroups**](sc_param.md#default-parameters)) is linked to the service.
+
+### find_servicegroup_in_list: parameters
+
+| parameter               | type   | optional | default value |
+| ----------------------- | ------ | -------- | ------------- |
+| a list of servicegroups | string | no       |               |
```
Comment (tanguyvda) [inline modules/docs/sc_event.md:741]:
```suggestion
| a coma separated list of servicegroups | string | no       |               |
```

File: modules/docs/sc_event.md [hunk 9]
```diff
@@ -739,19 +753,25 @@ The **find_servicegroup_in_list** method checks if one of the servicegroup in [*
...
 ## find_bv_in_list method
 
-The **find_bv_in_list** method checks if one of the BV in [**accepted_bvs**](sc_param.md#default-parameters) is linked to the BA.
+The **find_bv_in_list** method checks if one of the BV in the bvs list parameter ([**accepted_bvs or rejected_bvs**](sc_param.md#default-parameters)) is linked to the BA.
+
+### find_bv_in_list: parameters
+
+| parameter     | type   | optional | default value |
+|---------------| ------ | -------- | ------------- |
+| a list of bvs | string | no       |               |
```
Comment (tanguyvda) [inline modules/docs/sc_event.md:774]:
```suggestion
| a coma separated list of bvs | string | no       |               |
```

File: modules/docs/sc_event.md [hunk 10]
```diff
@@ -807,7 +827,13 @@ local result = test_event:is_valid_poller()
...
 ## find_poller_in_list method
 
-The **find_poller_in_list** method checks if one of the pollers in [**accepted_pollers**](sc_param.md#default-parameters) is monitoring the host.
+The **find_poller_in_list** method checks if one of the pollers in the pollers list parameter ([**accepted_pollers or rejected_pollers**](sc_param.md#default-parameters)) is monitoring the host.
+
+### find_poller_in_list: parameters
+
+| parameter         | type   | optional | default value |
+|-------------------| ------ | -------- | ------------- |
+| a list of pollers | string | no       |               |
```
Comment (tanguyvda) [inline modules/docs/sc_event.md:836]:
```suggestion
| a coma separated list of pollers | string | no       |               |
```

File: modules/docs/sc_event.md [hunk 11]
```diff
@@ -1039,6 +1065,39 @@ local result = test_event:is_valid_author()
...
 
+## find_author_in_list method
+
+The **find_author_in_list** method checks if one of the author in the authors list parameter ([**accepted_authors or rejected_authors**](sc_param.md#default-parameters)) is the author of a comment.
+
+### find_author_in_list: parameters
+
+| parameter         | type   | optional | default value |
+|-------------------| ------ | -------- | ------------- |
+| a list of authors | string | no       |               |
```
Comment (tanguyvda) [inline modules/docs/sc_event.md:1076]:
```suggestion
| a coma separated list of authors | string | no       |               |
```

**Review by tanguyvda — CHANGES_REQUESTED** [review]

**Review by tanguyvda — APPROVED** [review]

**Review by lucie-dubrunfaut — APPROVED** [review]

**Pull request merged**